lib/luminork-server/src/search/component.rs

```diff
@@ -0,0 +1,196 @@
+use std::{
+    collections::HashMap,
+    sync::Arc,
+};
+
+use futures::future::try_join_all;
+use itertools::Itertools as _;
+use serde::{
+    Deserialize,
+    Serialize,
+};
+use si_frontend_mv_types::{
+    index::change_set::ChangeSetMvIndexVersion,
+    reference::IndexReference,
+};
+use si_id::{
+    AttributeValueId,
+    ChangeSetId,
+    ComponentId,
+    WorkspacePk,
+};
+use telemetry::prelude::*;
+
+use crate::search::{
+    Error,
+    Result,
+    SearchQuery,
+    SearchTerm,
+};
+
+/// Search for components matching the given query in the given workspace and change set.
+pub async fn search(
+    frigg: &frigg::FriggStore,
+    workspace_id: WorkspacePk,
+    change_set_id: ChangeSetId,
+    query: &Arc<SearchQuery>,
+) -> Result<Vec<ComponentId>> {
+    // Grab the index first
+    let attribute_trees = attribute_tree_mv_index(frigg, workspace_id, change_set_id).await?;
+
+    // Spawn parallel fetch+match tasks for each component
+    let match_tasks = try_join_all(attribute_trees.map(|index_ref| {
+        tokio::spawn(match_component(
+            frigg.clone(),
+            workspace_id,
+            index_ref,
+            query.clone(),
+        ))
+    }))
+    .await?;
+
+    // Wait for them all to complete and collect the results
+    Ok(match_tasks.into_iter().flatten_ok().try_collect()?)
+}
+
+/// Fetch the AttributeTree MV for a component and match it against the query.
+#[instrument(level = "debug", skip_all, fields(id))]
+async fn match_component(
+    frigg: frigg::FriggStore,
+    workspace_pk: WorkspacePk,
+    index_ref: IndexReference,
+    query: Arc<SearchQuery>,
+) -> Result<Option<ComponentId>> {
+    let attribute_tree = attribute_tree_mv(frigg, workspace_pk, index_ref).await?;
+    if match_attribute_tree(&attribute_tree, &query) {
+        Ok(Some(attribute_tree.id))
+    } else {
+        Ok(None)
+    }
+}
+
+/// Match a component against a query or sub-query.
```

Comment:
This is the meat of where we process the actual query. If you have a query like `Instance region:us-east-1`, it will look like And(MatchValue("Instance"), MatchAttr("region", "us-east-1")). The And() processor first checks ithe component matches MatchValue("Instance"), and if it does, then it checks MatchAttr("region", "us-east-1").

If there are multiple attr:value matches, each MatchAttr will be processed separately, which means it will read through all the attributes and do the match separately. There might be ways to batch them and be a little more performant, but this is by far the most straightforward way to write it (especially when you take full boolean logic into account).